### ISSUE TITLE
Simple animation costs high CPU/GPU/Power usages (~20%) on iPhone 6

### ISSUE THREAD

**Issue by liyuqian** [flutter/flutter#31865]
See https://gist.github.com/liyuqian/67d848f651211613da0c1042905fe884

We should add a benchmark to devicelab and drive this number down.

Update: the benchmark is added, and we now have some clue on how to drive it down. Working on it.

**Comment by dnfield** [comment]
I think the bigger culprit here is GPU usage.  With the OpenGL backend, we seem to be pegged at 90% GPU utilization, which is really the big energy draw here.  With Metal, I see occasional jumps to that level but mostly staying much lower. This is another reason to move to Metal sooner than later.

/cc @chinmaygarde

**Comment by gaaclarke** [comment]
I was able to verify this inside of the profiler on device.  It looks like all that CPU time is related to syncing with the GPU (flush call).

![Screen Shot 2019-06-10 at 1 37 33 PM](https://user-images.githubusercontent.com/30870216/59225185-f2628e80-8b84-11e9-8149-7db48f2910ce.png)

**Comment by gaaclarke** [comment]
If you look at the backtrace it looks like our own timer is calling this code.  I wonder if the sync time would be less if we drove this code with a [CADisplayLink](https://developer.apple.com/documentation/quartzcore/cadisplaylink)?

**Comment by chinmaygarde** [comment]
We are using a [display link on iOS](https://github.com/flutter/engine/blob/master/shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm) already. The timer callback in the backtrace is because the fire callback is begin posted to the back of the UI task runner's message queue [here](https://github.com/flutter/engine/blob/master/shell/common/vsync_waiter.cc#L81).

**Comment by gaaclarke** [comment]
FWIW I ran the same code with the metal backend in release mode and it also it taking up 20%
![Screen Shot 2019-06-11 at 1 37 25 PM](https://user-images.githubusercontent.com/30870216/59305500-35893400-8c4f-11e9-82dc-9d798b58f9dc.jpg)

**Comment by dnfield** [comment]
I did not see CPU going down on Metal, I saw GPU going down.  Notice that the `Energy Impact` is now low, with little spikes.  If you run it on the OpenGL backend that energy usage remains at very high.  Going into energy impact should show you what is using energy - on Metal, it'll be CPU, on Open GL it'll be CPU and to a much greater extent GPU.

**Comment by liyuqian** [comment]
Depends on https://github.com/flutter/flutter/issues/31865

**Comment by liyuqian** [comment]
I tried a [simpler animation for the GPU (no fancy paths, just a rect)](https://gist.github.com/liyuqian/81def6f7d5b2126a7c21ae6b7c252c9c).

That has a CPU usage of 18%, which is about 3% lower than the 21% of our [circular progress animation](https://gist.github.com/liyuqian/67d848f651211613da0c1042905fe884).

I realized that 20% * 16ms = 3.2ms. From the observatory, it seems that we're spending about 1.6ms on the UI thread, and 1.6ms on the GPU thread (excluding the waiting of GPU driver work) no matter how simple the animation is.

On Android, the CPU usage (through `adb shell top`) is measured as a percentage over all cores. Hence for a 4-core device, a single-threaded app can at most have 25% usage. The 10% usage in that case actually translates to 40% usage if measured against a single core. Therefore, we actually had the same issue (or even worse since iPhones usually have better single-core performances) on Android, except that the number on Android are not very alarming.

Hence to significantly improve the CPU usage, we might need to some serious work. There might be no low-hanging fruit out there.

**Comment by dnfield** [comment]
What about the GPU usage though?  Anything that can be done to bring that down, short of going to Metal?

**Comment by dannyvalentesonos** [comment]
Thanks for looking into this. Having a very simple lottie animation running on a page in Release mode on an iPhone runs the phone very hot pretty quickly.
I noticed the same CPU usage @liyuqian mentioned on Android. I also noticed an android device getting warm when running a simple lottie animation as well, so that would explain why.

Something else I noticed yesterday running on an iPhone simulator in Debug mode. Having a very simple indeterminate `LinearProgressIndicator` widget on the page runs the CPU at 85%!

Not sure if you can reproduce that as well...

Thanks.

**Comment by liyuqian** [comment]
@dnfield : Yes, @gaaclarke discovered yesterday that we doing an additional full-screen texture compositing. That was used to make BackdropFilter faster on iOS. However, it makes other cases without BackdropFilter less efficient. @gaaclarke may be trying to remove that compositing if no BackdropFilter exists in the layer tree.

@dannyvalentesonos: thanks for your feedback! Adding CPU/GPU/Power usage tracking and improvement for real devices is a high-priority for us. On the other hand, the inefficiency of the iPhone simulator is a known issue to us as Flutter is only using software (CPU) backend without any GPU acceleration due to the simulator's bad OpenGL implementation. We understand that simulator's performance is critical to developers. However, the performance issue on the real iPhone devices is even more critical, hence we're not currently focusing on simulators right now due to resource limits. Please use a real iPhone during development if possible. Sorry for the inconvenience.

**Comment by dannyvalentesonos** [comment]
Thanks @liyuqian for the explanation. Running the following code in Release on a real iPhone 6 runs at 25% CPU with Very High Energy Impact.

```
@override
  Widget build(BuildContext context) {

    return Material(
        type: MaterialType.transparency,
        elevation: 0,
        child: Center(
            child: LinearProgressIndicator()
        )
    );
  }
```

**Comment by dannyvalentesonos** [comment]
@liyuqian I haven't. We work hard on making sure our build system can build with Flutter, and so we can't build the engine ourselves. Because of that, I have never spent time trying that. If it would be important for you that I do try, let me know and I can try and find some time to do so.

**Comment by liyuqian** [comment]
Don't worry about compiling engine that if you're not very interested. We'll soon add Metal to our default engine build so you don't have to compile. I suggested that just in case that you're so interested that you can't wait for us to include Metal by default.

**Comment by ckken** [comment]
same problem

**Comment by aleor** [comment]
> Don't worry about compiling engine if you're not very interested. We'll soon add Metal to our default engine build so you don't have to compile. I suggested that just in case that you're so interested that you can't wait for us to include Metal by default.

That sounds really great. Would you please share any estimations, when Metal is expected to become default?

**Comment by liyuqian** [comment]
@aleor : I guess that our official answer is always "it will be done when it's done" (the same answer you'd get if you asked the ETA of Flutter 1.0 release a year ago). CC @chinmaygarde to see if there's any estimation of the time point before which the Metal definitely won't be shipped.

**Comment by Hixie** [comment]
@liyuqian says this is blocked by #33899.

**Comment by Hixie** [comment]
This work is currently not expected to happen in the near future.

**Comment by dnfield** [comment]
I believe that migrating to metal would at least partially help this bug - which we are planning to do but tracking elsewhere.  Is declined customer request really right for this?

**Comment by Hixie** [comment]
Migrating to Metal may address this. Currently and for the near future we don't have anyone who has the bandwidth to actually study this specific issue directly, hence the milestone.

**Comment by eseidelGoogle** [comment]
I spoke with @cbracken just now.

My understanding of the problem statement from @dannyvalentesonos here is "Having a very simple lottie animation running on a page in Release mode on an iPhone runs the phone very hot pretty quickly.".  Which perhaps we should split out into a separate bug, as we potentially could solve that w/o solving this more general issue.

@cbracken's team is completely overloaded with a variety of Google-internal deliverables.  However, @timsneath has asked that we see if we could find someone to look into @dannyvalentesonos's concerns here.

We're going to see if @flar may be able (with a little help from @liyuqian) to look at this early next week and see if there is anything we can do to alleviate the shorter-term concerns.

My understanding from @cbracken (and @chinmaygarde) is that Metal support is probably only ~3 weeks of work away, but that that work is blocked behind other tasks at the moment.  Some parts of the system (e.g. PlatformViews) do not currently work with the metal backend.

@dannyvalentesonos please correct me if I've misunderstood your ask.  Please also feel welcome to file a new bug to better describe your short-term needs.

Thanks all.

**Comment by dannyvalentesonos** [comment]
@eseidel Thanks for your response.
A simple Lottie or Flare animation does run the phone very hot pretty quickly, but it is a combination of the fact that the CPU is running high for the entire length of the animation (which with Lottie/Flare could often be for as long as it is on the screen due to it looping), and the fact that without Metal, @dnfield mentioned the GPU thread is also pegged pretty high.

With Metal, I don't know how much lower the GPU thread will be running during animations, so I cannot really comment on that, but the high CPU usage for a very simple animation gets worse depending on the Lottie/Flare asset being used, which could cause frames to be dropped, especially when on a page that has a Material TextField, which seems to be slow to build/render on its own. Again, when switching to Metal, that could possibly get better, but I don't know.

Another thing I'd like to make sure is not being lost in this thread is that as @liyuqian pointed out, the CPU issue exists on Android as well, and might even be worse than iOS. Since that was mentioned, I did pay a little more attention to Android devices, now knowing that the CPU usage numbers was misleading me, and realized that devices such as a Nexus 9 do quickly run pretty hot with certain Lottie assets, although I don't think I've noticed any frames being dropped.

**Comment by eseidelGoogle** [comment]
I was asked also in a private thread if it would be possible to provide @dannyvalentesonos a Metal-enabled build for testing.  I suspect such would be relatively trivial for us to build a one-off and might be worth considering when someone takes a look at this bug next week.  Thanks!

**Comment by dnfield** [comment]
I sent an email about this before I saw the edit - if it helps I'm happy to help out with it.

**Comment by Hixie** [comment]
We're going to take a look at this this week to see if there are low hanging fruit; failing that our plan is to handle this primarily by moving to Metal, which we expect to do much later this year (see https://github.com/flutter/flutter/issues/18208).

**Comment by liyuqian** [comment]
We just landed a tool to automatically measure the iOS CPU and GPU usages: https://github.com/flutter/packages/pull/31

We will soon add a continuous CPU/GPU benchmark to our device lab through: https://github.com/flutter/flutter/pull/39439

@flar : any update on your side?

**Comment by flar** [comment]
I ran one of the simple linear progress indicator animations on an iPhone 6s for an hour and it wasn't very warm and only used 8% which equates to 12 hours of life. Considering that the screen was on the entire time, that is about average for screen energy usage for 6s.

**Comment by flar** [comment]
The impact of the "extra buffer copy" issue on iPhone is hard to gauge. At first glance it appears that it can drop us from 60fps to 30, but that turned out to be mostly a red herring. In fact, if you run the simple animation for a long time with the extra buffer enabled and disabled you can see that both have times when they run at 30fps and other times when they run at 60fps. It appears that some of our logic to decide whether to "render a frame ahead" gets stuck in a mode where we are always triggering an extra frame wait - so once we have any frame that didn't end in time we run a chance of encountering this condition. The condition is recoverable, but it requires another condition to come along and force us to miss a frame. We do need to fix that logic, but since it is mostly causing us to end up in resource lock waits it doesn't actually represent additional CPU/GPU usage. It just so happens that the condition is stochastically more likely to happen early on in the life cycle of an app when the extra buffer copy is enabled so if you start up the app and then look at its performance soon thereafter then you get a skewed picture.

Investigations still underway:

- What causes us to get stuck flushing a canvas and waiting for the next vsync randomly even when the total frame processing time is nowhere near the 16ms threshold (including overhead of receiving the vsync notifications)?
- Is the "high GPU and energy usage" tool really measuring actual energy usage?
   (Note that isn't to say that the original problem is not an issue, but we may have gone down the wrong path by thinking that the simple animations are showing the same problem.)

**Comment by flar** [comment]
I added a bunch of tracing calls throughout Skia and discovered that we are blocking in an actual GL call. On some frames the call takes < 1ms and we finish everything well within the vsync boundaries and then randomly some frames it will take ~15ms and finish just after we field the next vsync - possibly indicating that we are being blocked by the system for Vsync. More investigation will be needed to dig deeper into which semaphores or system calls are blocking us.

**Comment by flar** [comment]
I performed some empirical measurements of battery usage for various scenarios to get a real world measurement of the energy usage. In all cases, unless otherwise stated, I set the display to minimum brightness and disabled sleep so the screen would remain on. The tests were run for 4 hours to hopefully get statistically meaningful measurements of energy usage and some of the tests had to be repeated and I saw very similar battery drain, so the results should be fairly accurate.  Here are some results:

```
iPhone 6s, iOS 12.1.2, 4 hour runs
----------------------------------------------
Home screen (0%)   |************************                   (24%)
Stopwatch          |***********************|*******            (31%)
LinearProgress(1)  |***********************|****************   (40%)
LinearProgress(2)  |***********************|****************   (40%)
LinearProgress(3)  |***********************|*********          (33%)
Home screen (50%)  |***********************|***********        (35%)
--------------------------------------------------
(1) LinearProgress running on a recent build of Flutter,
    queue depth of 2,
    extra offscreen layer
(2) LinearProgress running on a recent build of Flutter,
    queue depth of 1,
    extra offscreen layer
(3) LinearProgress running on a recent build of Flutter,
    queue depth of 2,
    no offscreen layer
```
Note that using the iOS "energy impact" tool in Xcode, the (1) and (2) runs of LinearProgress measure as "Very High" energy usage, but the (3) run measures "Low" energy usage. The above give an indication of just how much energy counts as "Low vs Very High" with 9% usage over a static home screen indicating "Low" and 16% extra usage indicating "Very High". The (3) run is also quite close to the energy usage of the Stopwatch, a native iOS app written by Apple.

It bears noting that the "Home Screen" wasn't entirely static. At first I thought it was only redrawing once per minute for the clock in the notification bar, but then I realized that the clock icon was on the screen and it has an animating second hand so that state represents a once per second update rate. All other experiments included elements that updated on every vsync frame.

It appears that the queue depth of 2, which can frequently lock us into a state where we constantly have to wait until the next vsync to present the output to the screen, has no effect on the actual energy usage even though the Observatory traces appear ominous - in reality we are just seeing a kernel wait state in those traces and not actual CPU or GPU usage.

The extra offscreen layer used on iOS to improve BackdropFilter performance doesn't quite double the energy usage in terms of battery loss, but it causes the iOS tools to switch from "Low" to "Very High" usage and the app to use an extra 16% battery vs 9% (a 78% increase). The additional battery drain is smaller when taking the screen on-time into account (40% vs 33% for a 21% increase at 0% screen brightness). I'd need to run more tests to see how much increase is seen with a more realistic screen brightness, though.

At no point during these tests did the phone appear warm to the touch.

**Comment by liyuqian** [comment]
Great experiments @flar ! Can you please check if https://github.com/flutter/packages/tree/master/packages/gauge (only works with Xcode 10.2 or Xcode 10.3) can capture the difference between (1) and (3) using CPU/GPU percentages? If so, we can monitor the difference automatically in about 10 seconds, instead of having to manually monitor the battery after 4 hours...

**Comment by flar** [comment]
> Great experiments @flar ! Can you please check if https://github.com/flutter/packages/tree/master/packages/gauge (only works with Xcode 10.2 or Xcode 10.3) can capture the difference between (1) and (3) using CPU/GPU percentages? If so, we can monitor the difference automatically in about 10 seconds, instead of having to manually monitor the battery after 4 hours...

Here are the results:

```
       Test        Run   CPU     GPU
------------------|---|-------|-------
LinearProgress(1) | 1 | 13.8% | 19.2%
LinearProgress(1) | 2 | 15.4% | 20.7%
LinearProgress(1) | 3 | 15.6% | 20.3%
LinearProgress(1) | 4 | (gauge crashed)
LinearProgress(1) | 5 | 14.6% | 19.4%
LinearProgress(1) |avg| 14.85%| 19.9%
------------------|---|-------|-------
LinearProgress(3) | 1 | (gauge crashed)
LinearProgress(3) | 2 |  8.6% | 17.35%
LinearProgress(3) | 3 |  9.2% | 18.125%
LinearProgress(3) | 4 |  8.8% | 16.6%
LinearProgress(3) | 5 |  8.8% | 17.1%
LinearProgress(3) |avg|  8.85%| 17.3%
------------------|---|-------|-------
 ratio (3) / (1)  | % | 59.6% | 86.9%
```
If you factor both numbers together you get 59.6% * 85.9% == 51.8% total CPU+GPU usage.
From before, if we take the battery usage of the two scenarios and subtract the baseline we get 9% vs 16% which is a 56.25% savings, so those two numbers are fairly close especially when you consider the margin of measurement error for using single-digit percentage values of such small magnitude for the battery tests, the true battery usage ratio could have been +/- 6% or so (assuming 1 in 16 error).

**Comment by liyuqian** [comment]
Cool. It seems that we can verifiably save a significant amount of CPU and GPU energies by adopting (3) and removing the extra offscreen layer. Before all-in with this approach, one trade-off that we probably need to measure is the performance regression of backdrop filters on iOS without the extra offscreen layer. If it's Ok, we can remove the offscreen layer in all cases, otherwise we can only remove it without backdrop filters. (With Cupertino widgets, I'd expect the backdrop filter to happen a lot.)

Let's make it happen in Q4?

**Comment by dnfield** [comment]
FWIW, this lines up with why Metal seemed to perform better - Metal is not using an offscreen currently.

**Comment by flar** [comment]
We are working to get Metal builds available, hopefully soon. You can track the progress of the Metal project at https://github.com/orgs/flutter/projects/5

**Comment by liyuqian** [comment]
@flar : I still wonder if we can just remove the offscreen buffer on OpenGL. Skia may have improved the OpenGL a lot since we added that offscreen buffer workaround. Can you please confirm with ToT Flutter whether your BackdropFilter driver tests have a big performance difference on iOS with and without the offscreen buffer?

**Comment by flar** [comment]
> @flar : I still wonder if we can just remove the offscreen buffer on OpenGL. Skia may have improved the OpenGL a lot since we added that offscreen buffer workaround. Can you please confirm with ToT Flutter whether your BackdropFilter driver tests have a big performance difference on iOS with and without the offscreen buffer?

I ran the backdrop filter benchmark in the flutter framework repo on an iPhone 6s on iOS 12.1.2 and the average frame rasterization time went from 147ms to 555ms when I disabled the iOS offscreen surface. That's a 3.7x performance regression if we disable the offscreen surface. We should only do that conditionally if we do it.

**Comment by liyuqian** [comment]
Thanks for confirming that! Then it seems that we can only save the energy on simple scenes without `BackdropFilterLayer`. At least 50% energy save on the simple progress bar animation is still a huge win :smile:

**Comment by Hixie** [comment]
Our plan is to not look at this until Metal lands in a few months. I'm removing the "critical" label since we are not actually working on it. (Maybe we should just close it?)

**Comment by liyuqian** [comment]
@Hixie : Jim is actively working on this and he's planning to save 50% energy by Q4 on the simple progress bar animation. The solution is to simply remove the off-screen surface when there's no `BackdorpFilterLayer`.

**Comment by flar** [comment]
As mentioned in the comments of PR https://github.com/flutter/engine/pull/14104 that change should eliminate the extra offscreen surface used on many iOS apps. If the app uses any BackdropFilter widgets in its scene, then the offscreen surface will return and the savings will be thus negated. Eliminating the extra offscreen surface has been shown to improve the CPU/GPU usage of a simple app containing only a single LinearProgressIndicator by 50% and reducing the energy usage in the Xcode tools from "Very High" to "Low". Larger apps may see less of an improvement depending on the complexity of their animation and how much other content is rendered in the scenes that contain animations.

Android and web apps should not be affected by this fix.

Most notable with regard to using BackdropFilter widgets are a number of Cupertino widgets which use a BackdropFilter automatically on behalf of the application. Notably:

CupertinoNavigationBar - uses a BackdropFilter if the background color is not opaque
CupertinoTabBar - if bg color is not opaque
CupertinoPopupSurface - while popup is part of the scene
CupertinoSliverNavigationBarState - if bg color is not opaque
CupertinoActionSheet - while the dialog is part of the scene

A BackdropFilter is also used in the Material FlexibleSpaceBar widget, but only if the app manually specifies a StretchMode.blurBackground stretch mode in the constructor.

A BackdropFilter is used for a ModalRoute, but only if the app manually specifies a filter argument.

You can detect if a BackdropFilter is present in the scene using the `flutter run --verbose-system-logs` argument and look for lines in the output that look like `[VERBOSE0:compositor_context.cc(87)] Using SaveLayer to protect non-readback surface`

**Comment by liyuqian** [comment]
Thanks for the great work done by @flar , I've confirmed that the simple_animation benchmark in Flutter's device lab has a big improvement from 23% CPU and 14% GPU usages to  13% CPU and 8.5% GPU usage (see https://flutter-dashboard.appspot.com/benchmarks.html with filter `simple_animation`). And as Jim mentioned, in Xcode it goes from **very high energy** to **low energy** impact.

@dannyvalentesonos : can you please try the master branch tip-of-tree Flutter and see if the energy usage is improved on your app?

**Comment by dannyvalentesonos** [comment]
@liyuqian Do we need to build the engine, or will using the master flutter branch have these changes?

**Comment by liyuqian** [comment]
@dannyvalentesonos : you don't have to build the engine. The master flutter branch should have already included the newest engine needed.

**Comment by dannyvalentesonos** [comment]
@liyuqian I've just spent a few hours trying to get my environment ready for testing this and can't get it to work because of issues with the master channel at the moment.
```
Building flutter tool...
Could not find a file named "pubspec.yaml" in "/Users/<myusername>/.pub-cache/hosted/pub.dartlang.org/shelf_packages_handler-1.0.4".
Error: Unable to 'pub upgrade' flutter tool. Retrying in five seconds... (9 tries left)
Could not find a file named "pubspec.yaml" in "/Users/<myusername>/.pub-cache/hosted/pub.dartlang.org/shelf_packages_handler-1.0.4".
Error: Unable to 'pub upgrade' flutter tool. Retrying in five seconds... (8 tries left)
Could not find a file named "pubspec.yaml" in "/Users/<myusername>/.pub-cache/hosted/pub.dartlang.org/shelf_packages_handler-1.0.4".
Error: Unable to 'pub upgrade' flutter tool. Retrying in five seconds... (7 tries left)
Could not find a file named "pubspec.yaml" in "/Users/<myusername>/.pub-cache/hosted/pub.dartlang.org/shelf_packages_handler-1.0.4".
Error: Unable to 'pub upgrade' flutter tool. Retrying in five seconds... (6 tries left)
Could not find a file named "pubspec.yaml" in "/Users/<myusername>/.pub-cache/hosted/pub.dartlang.org/shelf_packages_handler-1.0.4".
Error: Unable to 'pub upgrade' flutter tool. Retrying in five seconds... (5 tries left)
Could not find a file named "pubspec.yaml" in "/Users/<myusername>/.pub-cache/hosted/pub.dartlang.org/shelf_packages_handler-1.0.4".
Error: Unable to 'pub upgrade' flutter tool. Retrying in five seconds... (4 tries left)
Could not find a file named "pubspec.yaml" in "/Users/<myusername>/.pub-cache/hosted/pub.dartlang.org/shelf_packages_handler-1.0.4".
Error: Unable to 'pub upgrade' flutter tool. Retrying in five seconds... (3 tries left)
Could not find a file named "pubspec.yaml" in "/Users/<myusername>/.pub-cache/hosted/pub.dartlang.org/shelf_packages_handler-1.0.4".
Error: Unable to 'pub upgrade' flutter tool. Retrying in five seconds... (2 tries left)
Could not find a file named "pubspec.yaml" in "/Users/<myusername>/.pub-cache/hosted/pub.dartlang.org/shelf_packages_handler-1.0.4".
Error: Unable to 'pub upgrade' flutter tool. Retrying in five seconds... (1 tries left)
```

I'll have to pick this back up in a little over 2 weeks unfortunately.

**Comment by dnfield** [comment]
@dannyvalentesonos you might have luck by just nuking your `.pub-cache` and letting it get rebuilt with a `pub get`.  Sounds like something in it has become damaged.

**Comment by dannyvalentesonos** [comment]
Thanks @dnfield That fixed my issue.

**Comment by dannyvalentesonos** [comment]
@liyuqian When running with the very latest master channel, I unfortunately still see the very same exact performance as before in our app.
I just ran the test with the 1.12 stable release and again with the master branch.
Stable release: 34%, Very High
master branch: 34%, Very High

**Comment by dnfield** [comment]
@dannyvalentesonos your app is very likely using a `BackdropFilter` somewhere.  The test case the current fix is optimizing for is for apps without a BDF.

@liyuqian @flar is there anything else being worked on to improve this for apps with BDF?

**Comment by liyuqian** [comment]
@dnfield : @flar is working on [ImageFiltered](https://github.com/flutter/engine/pull/14491) which could replace the `BackdropFilter` in many cases. 

@dannyvalentesonos : as @dnfield pointed out, it's very likely that your animation isn't that simple as it includes some `BackdropFilter`. https://github.com/flutter/flutter/issues/31865#issuecomment-564781986 listed some widgets that use `BackdropFilter`:
- CupertinoNavigationBar - uses a BackdropFilter if the background color is not opaque
- CupertinoTabBar - if bg color is not opaque
- CupertinoPopupSurface - while popup is part of the scene
- CupertinoSliverNavigationBarState - if bg color is not opaque
- CupertinoActionSheet - while the dialog is part of the scene

I wonder if you can share us a minimal Flutter app that reproduces your animation that has the 30%, Very High power usage? That way, we can optimize to your needs.

**Comment by flar** [comment]
> @dannyvalentesonos your app is very likely using a `BackdropFilter` somewhere. The test case the current fix is optimizing for is for apps without a BDF.

Run the app with --verbose-system-logs and look for output that indicates that it is having to use an offscreen layer due to screen readback. If so, then someone is using a BDF and blocking the fix.

Also note that Cupertino widgets sometimes install a BDF. @liyuqian just posted the list of cases I found by scanning the framework repo.

> @liyuqian @flar is there anything else being worked on to improve this for apps with BDF?

Just what @liyuqian already mentioned - if you are in control of the code that is adding the BDF over a background that you maintain then previously you were forced to use a BDF stacked over the background, but when we get ImageFiltered implemented you can wrap the background with the ImageFiltered and it can be blurred without requiring the base surface to be readable.  More details in [this issue](https://github.com/flutter/flutter/issues/13489) and the PRs that it links to.

**Comment by dannyvalentesonos** [comment]
I do not know how to pass --verbose-system-logs with the add2app scenario. Is there any documentation on how to do that? I have never found anything so far.
As for our scene, we are simply using a PageView, the video_player and a Flare widget. We, ourselves do not add anything that uses a BDF.

**Comment by dnfield** [comment]
Add `--verbose-logging` to the command line arguments in your Scheme.

**Comment by flar** [comment]
(`--verbose-system-logs`)

**Comment by dnfield** [comment]
@flar  that's the `flutter tool` switch for this - the actual shell argument is `--verbose-logging`, which is what you have to use if you're adding the command line argument directly in Xcode :)

**Comment by dannyvalentesonos** [comment]
I'll be very curious to see what comes of that, but it will have to wait until I'm back in 2 weeks.
Thanks guys.

**Comment by liyuqian** [comment]
You're welcome! Have a happy holiday.

**Comment by dannyvalentesonos** [comment]
@flar @liyuqian @dnfield 
I have added the `--verbose-logging` and this is all I'm getting
```
[VERBOSE0:gpu_surface_gl.cc(88)] Found 0 SkSL shaders; precompiled 0
[VERBOSE0:persistent_cache.cc(177)] PersistentCache::load failed: CAZAAAECA4AAAAAAAAABGAABAAOAAFAACUAAGAH7777777777777777777777777EAAAGAAAAAAAAAAAAAAAEAAAAAYAAWQA
[VERBOSE0:persistent_cache.cc(177)] PersistentCache::load failed: CAZAAAECAYAAAAAAAAAACAAAAAJQAAIADQABIAAPAALQAAAAAAABIAAYAAAAAAAAAAAAAAACAAAAAKAALIAA
[VERBOSE0:persistent_cache.cc(177)] PersistentCache::load failed: CAZAAAECAUAAAAAAAAABGAABAAOAAFAADQAAGAAQABOAAAAAAAAAAAAAAABAAAAAEAAFUAA
[VERBOSE0:persistent_cache.cc(177)] PersistentCache::load failed: CAZAAAICBEAAAAAAAAAEOAAAAAJQAAIA777777Y4AAKAB7777777777777777777EAAF2AAAAAAAA6IAAAACYAAEAAAAAAAAAAAAAAACAAAAAPAALIAA
[VERBOSE0:persistent_cache.cc(177)] PersistentCache::load failed: CAZACAICCEAAAAAAAAAGOAAAAAJQAAIA777777Y4AAKAAEYAAEAP777777777777EAAF2AAAAAAAABYAAAACYACUAAAAAAAAGQADMAABAAAAAAIAAAAAOAAAABCAAHQAAAAAAAAAAAAAAAB4AA6AAPAAHQAQAAAAA4AAAADAAA6QAAAAAAAHSAAAABWAABAAAAAAAAAAAAAAAAQAAAAHYAC2AA
[VERBOSE0:persistent_cache.cc(177)] PersistentCache::load failed: CAZAAAECAYAABEAAAAAAAAAAAAJQAAIA7777777777776GIAAAABQAAOAAAAAAAAAAAAAAACAAAAAKAALIAA
[VERBOSE0:persistent_cache.cc(177)] PersistentCache::load failed: CAZAAAECA4AAAAAAAAAEOAAAAAJQAAIA777777Y4AAKAB7777777777777777777EAAF2AAAAAAAAAAAAAAAEAAAAAYAAWQA
[VERBOSE0:persistent_cache.cc(177)] PersistentCache::load failed: CAZAAAACA4AAAAAAAIAACAAAAAAQAAAACMAACAA4AAKAADYAC4AAAAAAAAMAALIAAAAAAAAAAAAAAAQAAAACYAC2AA
[VERBOSE0:persistent_cache.cc(177)] PersistentCache::load failed: CAZAAAACA4AAAAAAAAAAEAAAAAJQAAIADQABIAAPAALQAAAAAAAAAAAAAAKAAGAAAAAAAAAAAAAAAAQAAAACYAC2AA
```

I'm running this in Release, so I don't know if that has anything to do with it. I'll try debug now.

**Comment by dannyvalentesonos** [comment]
The only logs I get in Debug are
```
flutter: Observatory listening on http://127.0.0.1:49951/8JG3v-riYwY=/
[VERBOSE0:gpu_surface_gl.cc(88)] Found 0 SkSL shaders; precompiled 0
```

**Comment by dannyvalentesonos** [comment]
I've also tested with the very simple flutter tree from https://gist.github.com/liyuqian/67d848f651211613da0c1042905fe884 and still getting 27% CPU and Very High energy without anything else running.

**Comment by liyuqian** [comment]
I'm quite surprised that even https://gist.github.com/liyuqian/67d848f651211613da0c1042905fe884
didn't have a significant decrease in CPU/GPU and energy use. That simple app shoudn't have a `BackdropFilterLayer`. @flar : Can you please try to reproduce the experiment and follow up with Danny while I'm out of office?

**Comment by flar** [comment]
I ran the circular progress indicator simple app from https://gist.github.com/liyuqian/67d848f651211613da0c1042905fe884 on my iPhone 6s in Profile mode under Xcode and also with the flutter gauge and it stayed at "Low" energy usage for over 5 minutes in Xcode with the CPU graph staying under 20% the whole time - usually around 18%. My flutter gauge readings were:

Run1: gpu: 8.4%, cpu: 17.125%
Run2: gpu: 8.8%, cpu: 17.475%
Run3: gpu: 9.8%, cpu: 17.424999999999997%

I also ran the linear progress indicator from https://github.com/flutter/flutter/issues/31865#issuecomment-504400303 in the same manner (iPhone 6s, Profile mode) and it also stayed "Low" for 5 minutes with the CPU graph under 18% the entire time - lots of numbers in the 15-17% range. The flutter gauge readings I got for that app were:

Run1: gpu: 8.6%, cpu: 15.3%
Run2: gpu: 8.0%, cpu: 16.1%
Run3: gpu: 8.6%, cpu: 16.0%

**Comment by flar** [comment]
One slight hitch, when I went to check which commit hash I was running off of I noticed that I had a small patch applied to diagnose another issue that shouldn't be related. Indeed I repeated the experiments without the patch using https://github.com/flutter/flutter/commit/525855d5954a18562d53466f252a7b9f4f48f487 and got similar results to what I had before. The run was Low for 5 minutes except for 2 instances where it crept over the line to the bottom of the yellow "High" range for one sampling period. That may have happened last time as I was spending a lot more time watching the CPU graph before, but it stayed solidly green for the last 2 minutes of the run and the CPU graph stayed well below 20% - with a lot more numbers in the 15-17% range. The 3 readings I got from the flutter gauge for this official commit were:

Run1: gpu: 8.2%, cpu: 18.325%
Run2: gpu: 8.6%, cpu: 18.35%
Run3: gpu: 9.6%, cpu: 18.825%

**Comment by liyuqian** [comment]
@dannyvalentesonos : it seems that @flar couldn't reproduce your 27% CPU and Very High energy result on https://gist.github.com/liyuqian/67d848f651211613da0c1042905fe884 . Shall we schedule a video conference to figure out why we couldn't reproduce your energy issue?

**Comment by dannyvalentesonos** [comment]
@liyuqian Sure, that could be good.

**Comment by flar** [comment]
Some differences in environment identified during the video call:

- @dannyvalentesonos is using an iPhone 6, but I am using an iPhone 6s (80% faster processor apparently and double the RAM)
- @dannyvalentesonos has been testing under add-to-app, whereas I was running the app directly

We had @dannyvalentesonos run the app directly (not using add-to-app) on his iPhone 6 on the latest version of Flutter and the energy usage dropped dramatically, but not as much as I saw on my iPhone 6s testbed. Percentage-wise the drop looked about the same on the energy usage graph, but both the before and after figures on his iPhone 6 device were higher than the equivalent figures I saw on my 6s.

The next step is to create a simple add-to-app wrapped version of these test cases and compare the energy usage performance of that app.

**Comment by liyuqian** [comment]
Our ability to add an add-to-app benchmark for this is currently blocked by https://github.com/flutter/flutter/issues/45435 . Will resume later.

**Comment by liyuqian** [comment]
@flar : #45435 is now closed so this should be unblocked. However, the bad news is that the simple_animation_perf on iPhone 6s has ~25% CPU usage with OpenGL and ~20% CPU usage with Metal for Flutter 77b4505c8014330fba2f2706c658a64bc5d1891b. I believe that's the number before your optimization in https://github.com/flutter/engine/pull/14104 landed. According to https://flutter-dashboard.appspot.com/benchmarks.html, the CPU usage dropped from ~23% to ~14% after the optimization. Can you please check if that optimization is still working properly now, or if there's some uncaught regression?

**Comment by flar** [comment]
My optimization in https://github.com/flutter/engine/pull/14104 was for a benchmark that added and removed a BDF. The simple_animation_perf is a completely different animation which may have different performance, so i don't think that numbers I was using as before/after on the post_backdrop benchmark can be compared here.

**Comment by flar** [comment]
In looking back on the code that used gauge to measure the CPU/GPU usage, the measurement tool was run after the benchmark completed. The new code that uses metrics from the engine traces now takes those measurements as the benchmark is run. If you look at the `total_cpu_usage` in the trace log, the first few samples included are quite a bit higher than the rest. This could account for much of the difference in the measurements.

**Comment by liyuqian** [comment]
@flar : in https://github.com/flutter/engine/pull/14104#issuecomment-561931914 you mentioned "LinearProgressApp" which is essentially the same as "simple_animation_perf". I believe your optimization is about removing a costly offscreen surface when there's no BDF, and the LinearProgressApp/simple_animation_perf doesn't have such BDF.

Ah, the initial high cost of CPU usage could explain the difference. Then I shall probably modify the test to only measure the CPU usage after the animation stabilizes so we're not measuring those initial spikes.

**Comment by flar** [comment]
I see the numbers now. Yes, that was a comparable benchmark (though standalone, not in the context of the macrobenchmark wrapper in which we run the simple_animation_perf).

Ignoring the first few measurements may not bring the CPU usage all the way down to parity with that chart from https://github.com/flutter/engine/pull/14104#issuecomment-561931914 but the energy usage of the linear_progress app was shown as Low when I ran it just now with Xcode. I didn't try running the macrobenchmark/simple_animation page under Xcode, though.

**Comment by liyuqian** [comment]
@flar : can we close this now as
- Simple animations now have low energy usage reported by Xcode. See recent experiment in https://github.com/flutter/flutter/issues/64205#issuecomment-707426617.
- We still see 5% CPU usage decrease today from optimizations in https://github.com/flutter/engine/pull/14104#issuecomment-561931914.
- [GPU usage is still quite low at ~8%][1] today (down from ~16% before optimizations in https://github.com/flutter/engine/pull/14104).

Or do you have more work to do for add-to-app regarding this? We may also create a dedicated issue for add-to-app so we don't have to spam this mega issue.

[1]: https://flutter-flutter-perf.skia.org/e/?begin=1602801424&end=1603255619&queries=sub_result%3Daverage_gpu_usage%26test%3Dsimple_animation_perf_ios&requestType=0

**Comment by flar** [comment]
Closing per recommendation by @liyuqian

**Comment by ckken** [comment]
OH YEAH！
